### DISCUSSION TITLE
How to enable compression (gzip/brotil) in nextcloud-aio?

### DISCUSSION THREAD

**Discussion by SubhashBose** [nextcloud/all-in-one#2574] [Ideas] [answered]
I want webserver traffic compression for Nextcloud-AIO, but I couldn't find any option to enable this. Is it possible to do this in NC-AIO?

For my setup, I see that NC-AIO sends uncompressed traffic for the UI of ~32MiB but on compression, it can be just ~8MB. That's about 75% compression. 
The majority of the compressible traffic is from static assets, like JS and CSS files. The worst is for core-common.js file which is 13MB, but on compression, it is just ~2MB.
So, I think for NC setup compression is very important for having a fast Web-UI loading. I was hoping NC-AIO does compression by default, but that's not the case.

One option can be putting a reverse proxy in between which will handle the compression, but this is not what I want. Nextcloud AIO already runs Apache and Caddy within dockers, and I just don't want to add another webserver in between. Best would be to have the option to enable compression in Apache or caddy which is already included.

**Comment by Zoey2936** [comment]
Seems like adding: `encode zstd gzip` to the caddyfile would be enough to do this (https://caddyserver.com/docs/caddyfile/directives/encode), brotli is not supported by caddy by default. I think enabling gzip in apache (not caddy) is also possible, but I think caddy would be enough. @szaimen what do you think about this?

**Reply by SubhashBose** [reply]
Thanks, this is what I am looking for. In apache, mod_gzip has to be enable to get this. But as you said, for caddy it is just as simple as adding this one line to caddyfile, then no need for apache compression. I hope this is implemented in NC-AIO.

Additionally, if you want, you can add a optional flag to initial docker run to disable compression if some user wants to.

**Comment by szaimen** [comment]
`encode zstd gzip` sounds good in general but I would probably only enable it for 
![image](https://github.com/nextcloud/all-in-one/assets/42591237/eab59bd4-ed3e-4732-8f3e-c7b15850bab5)
which seems to be the main problem here to not cause any unexpected sideeffects. 

Also @Zoey2936 would this work behind reverse proxies or would we need to disable that for them?

**Reply by Zoey2936** [reply]
From my experience, compression done by an origin and additional compression by the reverse proxy is no problem

**Comment by szaimen** [comment, marked as answer]
This will be fixed with https://github.com/nextcloud/all-in-one/pull/2576

**Reply by szaimen** [reply]
This is now released with v6.0.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Comment by Zoey2936** [comment]
maybe `encode zstd gzip` should still be added to the caddyfile of the mastercontainer and apache, to also compress content from collabora and other proxied content. https://caddyserver.com/docs/caddyfile/directives/encode

**Reply by szaimen** [reply]
I see. But then we should probably disable brotli again to not double-compress it? Also downside is that we cannot limit it for certain file formats, right?

**Reply by Zoey2936** [reply]
you can limit it (as shown at the end of caddy encode docs, but the default looks good) and double compression should not happen

**Reply by szaimen** [reply]
Right, did not see this... Then maybe lets disable brotli again and enable gzip for the same file formats but inside Caddy. Can you create a PR? :)

**Reply by Zoey2936** [reply]
when I have time yes
